### PR TITLE
feat(hex): support update-lockfile strategy

### LINKED ISSUE #13618: supporting rangeStrategy=update-lockfile with mix manager
### What would you like Renovate to be able to do?

We use ranges in our `mix.exs` and then have specific versions in `mix.lock`. We would live renovate to update `mix.lock` when a newer version is available (currently it only updates when `mix.exs` needs to be updated).

See https://github.com/renovatebot/renovate/discussions/13616#discussioncomment-1988704 for more details

### If you have any ideas on how this should be implemented, please tell us here.

Something similar to https://github.com/renovatebot/renovate/pull/3108

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by pdm-jd**
## Changes

* Support `rangeStrategy=update-lockfile` for the `mix` manager
* Fixed fixture `mix.exs` to have a realistic structure - no duplicate dependencies with different versions

## Context

Closes #13618 

Depends on https://github.com/renovatebot/renovate/pull/27924

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

It looks like the documentation for [Hex versioning](https://docs.renovatebot.com/modules/versioning/#hex-versioning) should automatically update to include `update-lockfile`, based on `tools/docs/versioning.ts`

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Example PRs created with this code for these dependencies https://github.com/pdm-jd/renovate-mix-tests/pulls

```elixir
  defp deps do
    [
      {:tesla, "~> 1.4"},
      {:poison, "~> 4.0"}
    ]
  end
```

image

### CODE REVIEW

File: lib/modules/manager/mix/__fixtures__/mix.exs [hunk 1]
```diff
@@ -20,18 +20,18 @@ defmodule MyProject.MixProject do
 #   {:broadway_dashboard, "~> 0.2.2"},
       #    {:broadway_dashboard, "~> 0.2.2"},
       {:postgrex, "~> 0.8.1"}, #  {:broadway_dashboard, "~> 0.2.2"},
-      {:ecto, ">2.1.0 or <=3.0.0"},
+      {:foo_bar, ">2.1.0 or <=3.0.0"},
```
Comment (viceice) [inline lib/modules/manager/mix/__fixtures__/mix.exs:23]: why rename deps?

**Review by viceice — COMMENTED** [review]
please split this PR in a manager and a versioning part.

File: lib/modules/manager/mix/__fixtures__/mix.exs (hunk 1 above)
Comment (pdm-jd) [reply lib/modules/manager/mix/__fixtures__/mix.exs:23]: This fixture is an invalid `mix.exs` file.

Here's a cleaned version of the original (removed fake private packages and updated the tag for `cowboy` to a real one). The screenshot below shows the output when trying to run  `mix deps.get` - it doesn't work because of duplicated dependencies. I thought it would be important to have the fixture reflect a real life example.

```elixir
defmodule MyProject.MixProject do
  use Mix.Project

  def project do
    [
      app: :my_project,
      version: "0.1.0",
      elixir: "~> 1.14",
      start_permanent: Mix.env() == :prod,
      deps: deps()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps() do
    [ #{:broadway_dashboard, "~> 0.2.2"},
#{:broadway_dashboard, "~> 0.2.2"},
#   {:broadway_dashboard, "~> 0.2.2"},
      #    {:broadway_dashboard, "~> 0.2.2"},
      {:postgrex, "~> 0.8.1"}, #  {:broadway_dashboard, "~> 0.2.2"},
      {:ecto, ">2.1.0 or <=3.0.0"},
      {:cowboy, github: "ninenines/cowboy", tag: "2.11.0"},
      {:phoenix, git: "https://github.com/phoenixframework/phoenix.git", branch: "main"},
      {:ecto, github: "elixir-ecto/ecto", ref: "795036d997c7503b21fb64d6bf1a89b83c44f2b5"},
      # {:secret, "~> 1.0", organization: "acme"},
      # {:also_secret, "~> 1.0", only: [:dev, :test], organization: "acme", runtime: false},
      {:ex_doc, ">2.1.0 and <=3.0.0"},
      {:jason, ">= 1.0.0"},
      {:jason, "~> 1.0",
        optional: true},
      {:hammer_backend_redis, "~> 6.1"},
      {:phoenix, "== 1.6.14"}
    ]
  end
end
```


image

**Comment by pdm-jd** [comment]
> please split this PR in a manager and a versioning part.

@viceice to confirm, would you like me to close this one and submit two PRs that reference this one?

**Comment by viceice** [comment]
no, just create one new PR with the changes which need to be merged first. currently don't see if it's versioning or manager, so decide yourself.

**Review by viceice — COMMENTED** [review]
blocked by #27924

**Comment by pdm-jd** [comment]
This is ready for re-review. I've validated it still works as expected with the code.

Reference: https://github.com/pdm-jd/renovate-mix-tests/pulls

image

File: lib/util/pretty-time.spec.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ describe('util/pretty-time', () => {
   });
 
   describe('satisfiesDateRange()', () => {
-    const t0 = DateTime.fromISO('2023-07-07T12:00:00');
+    const t0 = DateTime.fromISO('2023-07-07T12:00:00Z');
```
Comment (viceice) [inline lib/util/pretty-time.spec.ts:50]: why this change?

Comment (pdm-jd) [reply lib/util/pretty-time.spec.ts]: This test was failing for me locally, but it doesn't seem to pass in CI. I've removed it.

**Review by viceice — APPROVED** [review]
